### PR TITLE
Add new tutorial on managing software

### PR DESCRIPTION
**Pull request by s-makin**
This tutorial, aimed at newer Ubuntu users, should provide a comprehensive walkthrough of
managing software, which explains many of the terms and concepts they need to know
to use the rest of the content of the server docs and help them be successful.

Note: I'm having issues with Multipass currently, so the later part of the tutorial I
haven't been able to properly test to ensure that it works on a fresh VM

### CODE REVIEW

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:154]: You could also say they can press Enter for the default (Yes in this case because Y is capitalized)

I also think you should have this section start with "This command will ask you to confirm..." upfront. Not having that information be the first thing that came after the `upgrade` command is confusing. Or, you could inform the user before they run it that they'll be asked a prompt

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:254]: Should you give a quick mention what man pages are?

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:263]: This part (implicitly) guides the user to actually run this command though. But if they install Apache here, then they won't be able to follow the next section, which assumes they don't have Apache installed yet.

Can you pick a different package for these two examples with apache?

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:277]: This section made me think - maybe you should show how to install and uninstall packages right at the beginning of the tutorial? So the user is equipped with that knowledge in case they get off track in your tutorial. It could be done as a quick overview of some commands they might need along the way (or something like that)

There's a lot of picky steps here, and if you accidentally install something you shouldn't and get off, it's difficult to get back (well if your user isn't that familiar with Linux at least, which might be the case for this tutorial)

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:362]: Maybe include a step to show that ssl-cert is already installed? (`apt list ssl-cert`?)

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:372]: Can you note how to uninstall apache2 if they accidentally already clicked Y? Otherwise, they won’t be able to follow any of this part of the tutorial

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:626]: Doesn’t it prompt them if they want to override their existing conf files though? Do you want to note that? (I’m not actually sure if they’re always prompted–Landscape Server prompts if they want to overwrite configuration files)

Comment (YanisaHS) [inline tutorial/managing-software.md:626]: I think it'd be good to provide an example of information that could be in conf files. The user is really thrust into conffile world without much context here (aside from the official deb doc)

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:660]: Can you explain what the parts of this command are doing? So it’s not like “hello beginner just run this magic”

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:711]: After running this, I got `Failed to restart apache2.service: Unit apache2.service not found.`

I don't think the tutorial actually has us reinstall apache2, so that should be added (unless I just missed it)

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:787]: I think you should put the expected output before this.

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:796]: I like this, but I think you should have a little bit of this at the top in case the user accidentally installs something and gets off track from your tutorial. You could still have this section here but reword it slightly

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:1110]:
```suggestion
Every Ubuntu release (`noble`, `jammy`, etc) is also split into **pockets** (sections of packages)
which are related to the development/release lifecycle:
```

Or similar? Your doc doesn't really note what pockets are. It defines them officially, but it doesn't clear up what they actually are to a beginner

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:1232]: why do we want to configure the source tree? What is the source tree and why does it need to be configured? (You lightly explain below, but I think context should be provided before they run the command)

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:1240]: The user is thrown into make/Makefiles here too. Can you provide context?

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:1260]: I’m not a fan of using Ruby as the example for a couple reasons:
1. Ruby isn’t very popular anymore, so some users (particularly younger ones who weren’t connected to tech during the Ruby on Rails time) might not be familiar with it, and a more general example would be less distracting
2. Ruby’s package manager is called with `gem`, but Ruby also calls its packages “gems”, so it’s difficult for people brand new to this stuff to even Google what “ruby” or “gem” is

You could use an example with `pip` instead?

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:1258]: Prompts you to Y/N to continue

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:1259]:
```suggestion
sudo gem install fpm
```

File: tutorial/managing-software.md
Comment (YanisaHS) [inline tutorial/managing-software.md:1265]: can you give context to what’s in this command? (say what all the flags are/etc).

I also got this error:
```bash
Invalid package configuration: Cannot package the path './usr/local/apache2', does it exist? {:level=>:error}
```

**Review by YanisaHS — APPROVED** [review]
This tutorial is insane but also awesome!!! My review is done :hot_face: 

("Approving" because I accidentally made this a formal review instead of just single comments ...)

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:14]: Added a prerequisites section :)

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:27]: I'm wrestling with this a bit. It doesn't feel like the right place to talk about it, at this point the sudo is just part of the vehicle to get to the tutorial, and there are no good up-to-date or reliable resources to point to. I might find a place to drop in a note later on -- someone who's not a complete beginner will know about sudo already, and a complete beginner might not even realise the significance. Or I could sidestep it by giving a link to the snapstore page in the instructions.

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:48]: Great point! Added

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:79]: I'm not sure that `pip` use is extensive/widespread enough to make the comparison meaningful (especially because APT is the default in Ubuntu), but I can/should probably make the package management aspect more explicit. I'll have a fiddle and see what improvements I can make.

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:154]: Reorganised the contents a bit to hopefully make the "you need to confirm" part a bit more visible, and added the Enter tip (I didn't know about that one!)

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:254]: Hmm...yes, probably! And also a link to the online ones. Thanks for the reminder :)

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:277]: I think that's a great point. I don't think my "we won't install anything yet" is strong enough given that a user may just start running commands on autopilot. I think I'll generalise these commands so that they just won't install anything.

Comment (s-makin) [reply tutorial/managing-software.md:277]: On reflection, this also seemed like the ideal time to talk about sudo, so added that content here

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:362]: If they're following the commands/instructions exactly, a user shouldn't have ssl-cert installed yet. This whole section is just about the different dependency types and how to find out what's recommended/required/suggested etc (and a bit of an introduction in how to use flags).

Comment (s-makin) [reply tutorial/managing-software.md:362]: I need to make it a bit stronger in the intro to the section I think.

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:626]: Good point, I missed that. I've added a bit on what sort of info *might* be in a package (since it depends on the package).

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:660]: Not easily/in a way that doesn't invite more questions, but I've tried :upside_down_face:. I've added the bits of the command to the relevant keywords of the paragraph above to highlight them, but it's still not ideal.

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:711]: Do you know, you're completely right. In all the previous steps I forgot to add that in! Great catch

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:660]: I've also added the manpage command, partly so users can investigate the options if they want, but also to get them used to using the manpages. I'm still not 100% happy, but I think it's better than what I originally had.

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:711]: Fixed

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:787]: That's fair! :) added

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:1110]: This is where everything has the potential to become very confusing lol. The thing that really is "sections" of packages is the components, where they're clearly divided into one or the other. You can't have a package in both main and universe afaik. Pockets aren't sections of packages, so much as sections of components. I've never been able to find a good, non-confusing explanation, so I'll try and write one.

Comment (s-makin) [reply tutorial/managing-software.md:1110]: I've added a mermaid diagram as well, this may help a bit

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:1260]: In general, I'm ok with providing exposure to older tech (we even have content about Ruby in the server docs). In this case, we can't get `fpm` through `pip` so I'm a bit tied to Ruby here :/ (the alternatives you can get through pip are even more fiddly and annoying to use, so this is the least-worst option imo)

File: tutorial/managing-software.md
Comment (YanisaHS) [reply tutorial/managing-software.md]: FWIW I described pockets as "sections" and components as "categories" and made a diagram here: https://ubuntu.com/landscape/docs/explanation-about-repository-mirroring. It's def still confusing, but just passing the link along if it helps to see a second opinion on how that stuff could be explained :blush:

File: tutorial/managing-software.md
Comment (s-makin) [reply tutorial/managing-software.md:1386]: Context added!

File: tutorial/managing-software.rst
Comment (cpaelzer) [reply tutorial/managing-software.rst:1333]: If someone wants to do the very same BUT with also automatically handling dependencies to be resolved from the archive, they can use `apt install ./filename`. You need leading / or ./ to make sure apt knows it is not a package name from the repo. But that allows to install a .deb with the simplicity of if it has dependencies that can be resolved - they would.

Therefore I'd prefer to suggest that. WDYT?

File: tutorial/managing-software.rst
Comment (cpaelzer) [reply tutorial/managing-software.rst:1335]: Furthermore this whole topic is so very different based on which software you talk about - that none of the instructions might match when someone does something else. So even when moving elsewhere, you should add even more intense "this will be very different depending on what you do, this is just one example"

And again, not good to be here in the tutorial.

Makes the otherwise great "how you should use it" complicated end ends with showing something people should better not use - at least not until they really++ know what they do.

Moving installing from manually downloaded .deb
And building source to separate pages each and linking from the tutorial as "there is more, but you should almost never need that and it comes with drawbacks to maintainability of the systems (no updates, ...)" is what I'd expect.

File: tutorial/managing-software.rst
Comment (cpaelzer) [inline tutorial/managing-software.rst:1607]: You did not mention the build from source here, which is good - I do not want it here :-)

But this list also forgets about the install from a .deb file - which ok'ish.
I'm unsure if that should be here or also in a page of its own with big disclaimers (essentially the third party explanation we have).

**Review by cpaelzer — CHANGES_REQUESTED** [review]
Already quite good, I've left hopefully helpful suggestions.
But even more so at the end there was some content that I disagree (it should not be part of the tutorial).

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:8]: In a how-to I would, but here I don't want to encourage anyone to skip ahead over the setting up a VM part.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:68]: Definitely worth thinking about. I think if we wanted to expand the tutorial with a more cloud-based topic that might make a really good extension to it. For now I wanted to try and give a soft landing just to introduce the terminology and (hopefully) make clear that there's not a massive difference between Server/Desktop. That was the thinking behind the repetition anyway. I know a lot of Desktop users are likely to end up using this tutorial, so I want them to feel "safe" to do so.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:78]: Good idea, I've removed it and just left "Welcome to Ubuntu"

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:93]: That's much better, thanks!

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:94]: went with "pre-installed" - thanks

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:93]: On reflection I shortened it to "is to make sure we have the latest versions..." and avoid both update and upgrade.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:96]: I'm happy with that - changed to your suggestion

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:104]: I take your point - but that might be a complicated way to describe it. I've tried to come up with an alternative to the original that steers a bit closer to APT just being an interface. Let me know if this is correct or if it still needs a bit more work.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:210]: Much better. I've also added a parenthetical example to hopefully make it even clearer what we're talking about.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:227]: Since we're using APT to search, I've updated the section heading to "Searching with APT" (hopefully accurate enough to reflect what you're suggesting here) and updated the text with your suggested wording.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:303]: I'm not sure. It seems like a helpful thing to mention, but since most guides and instructions a new user runs across will show them sudo commands, my intention was to talk about sudo and encourage users to slow down and read and understand what's presented (and not just spam "enter" through confirmation screens). I don't think I've ever seen the simulate flag before, so it's not commonly used in documentation. I think I'll add it, but I propose to do it at the end of the section. wdyt?

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:346]: I've tried to reword this a bit without going too far into explaining dependencies. Let me know if it's better or not.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:391]: Made the wording a bit stronger in both places

Comment (s-makin) [reply tutorial/managing-software.rst:391]: (in the second place, made into a warning)

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:577]: Agreed and good point - added a little explanation here to that effect. I'm still pondering on where to introduce the --dry-run option. This doesn't feel like the most relevant place for it.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:703]: I think that's much better, updated to that.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:704]: Good catch - made it more specific

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:720]: That's a great point. I picked "default" because it made the table more compact but I see the ambiguity I've created by doing that. I've changed it to "changed/not changed by maintainer" -- hopefully this is better!

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:1275]: I didn't realise it would split that, never seen that happen before. I've tried to reword the paragraph a bit to avoid that word appearing at the end of the line, but it will depend on the screen size to some extent. At least on a regular monitor it shouldn't happen anymore.

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:1335]: Removed this entire section, as discussed

File: tutorial/managing-software.rst
Comment (s-makin) [reply tutorial/managing-software.rst:1607]: I don't want to draw too much attention to it, since we prefer people don't do it and are only showing it as an FYI because we're doing it in a sandbox. Actually, now that I reflect more on the reasons to remove all the building from source stuff, I wonder if we should also remove the install from .deb stuff - but it does provide a convenient segue into snaps.

**Pull request merged**
